tornadox/blocks/step_impl/first_order/ek1_projmatfree_d_nu.py

```diff
@@ -0,0 +1,147 @@
+"""Projection-matrix-free version of the EK1."""
+
+
+from functools import partial
+
+import jax
+import jax.numpy as jnp
+
+from tornadox.blocks import sqrtutil
+
+
+@partial(jax.jit, static_argnames=("f", "df", "num_derivatives"))
+def attempt_step_forward_only(
+    *, f, df, m, c_sqrtm, p, p_inv, a, q_sqrtm, num_derivatives
+):
+    """Attempt a full step.
+
+    Extrapolate-then-correct.
+    Includes error estimation.
+    Includes a scalar, time-varying diffusion.
+
+    Assumes a dimension-derivative ordering.
+    """
+
+    # Apply the pre-conditioner
+    m, c_sqrtm = p_inv * m, p_inv[:, None] * c_sqrtm
+
+    # Predict mean and compute error
+    m_ext = p * (a @ m)
+
+    # Linearise ODE and estimate error
+    fx, Jx, m_obs = _linearise_ode_projmatfree(
+        f=f, df=df, m_ext=m_ext, num_derivatives=num_derivatives
+    )
+
+    err, diff_sqrtm = _estimate_error_projmatfree(
+        Jx=Jx,
+        m_res=m_obs,
+        q_sqrtm=p[:, None] * q_sqrtm,
+        num_derivatives=num_derivatives,
+    )
+
+    # The full extrapolation:
```

Comment:
Would it be helpful to include something like "predict covariance" somewhere, if that's what's happening here?

Reply:
I prefer to use "extrapolate" and "correct", as I think that "predict" and "update" are a bit overloaded in this world of algorithms. But I will try and improve the comments a little bit :)

Reply:
Now that the first one also says "extrapolate" and not "predict" anymore I think it's fine, thanks :)